src/icon-button.tsx

```diff
@@ -356,4 +369,52 @@ export const BaseIconButton = (props: IconButtonProps): JSX.Element => {
     );
 };
 
-export default IconButton;
+type ToggleStateProps = {
+    Icon: React.FC<IconProps>;
+    type?: IconButtonType;
+    backgroundType?: IconButtonBackgroundType;
+} & AriaProps;
+
+interface BaseToggleProps {
+    checkedProps: ToggleStateProps;
+    uncheckedProps: ToggleStateProps;
+    onChange?: (checked: boolean) => void | undefined | Promise<void>;
+    checked?: boolean;
+    defaultChecked?: boolean;
+}
+
+export type ToggleIconButtonProps = BaseProps & BaseToggleProps;
+
+export const InternalToggleIconButton = React.forwardRef<
+    TouchableElement,
+    ToggleIconButtonProps & {isOverMedia?: boolean}
+>(({checked, defaultChecked, checkedProps, uncheckedProps, onChange, ...props}, ref) => {
+    const [checkedState, setCheckedState] = React.useState(!!defaultChecked);
+
+    const handleChange = () => {
+        if (checked === undefined) {
+            // if onChange is asynchronous, wait until it finishes and change the state if there was no error
+            const result = onChange?.(!checkedState);
+            if (result) {
+                return result.then(() => setCheckedState(!checkedState));
+            } else {
+                setCheckedState(!checkedState);
+            }
```

Comment:
I think this should be equivalent
```suggestion
            Promise.resolve(result).then(() => setCheckedState(!checkedState));
```

Reply:
Discussed offline. This is not equivalent. In my code, the else statement returns void. In yours, it returns a promise.

This affects the RawIconButton component, because it checks if the result of calling onPress is a promise, and performs a setState if that is the case. We don't want that.